docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,252 @@
+```yaml
+- apiGroups:
+  - resources.azure.com
+  resources:
+  - resourcegroups
+  - resourcegroups/status
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
+The above permissions may overlap with ASO resources already installed by a user not intended to be managed by CAPZ.
+
+### Risks and Mitigations
+
+The most immediate risks with adopting ASO are regressions of various forms. Firstly, Azure API calls are core to CAPZ's functionality, so even small behavioral differences may cause issues preventing Azure resources from being reconciled as expected. Next, the modified reconciliation flow may cause operations against workload clusters to take longer to complete, in particular if it takes more reconciliations to perform a task. Finally, integration with ASO may introduce new or worsen existing bottlenecks limiting how many CAPZ resources can be managed. Scaling bottlenecks might be Azure API rate limits or management cluster Kubernetes API server load because of the increased number of Kubernetes resources. These risks will be monitored closely as described in the [Test Plan](#test-plan). Any significant regressions found can be rolled back backwards-compatibly by reverting the relevant changes.
+
+Along those same lines, there is a risk that if a service is implemented both with the SDK and with ASO in different supported versions of CAPZ, then the fix for an issue affecting both versions will need to be implemented separately for each version. Such issues are not expected to be common given the implementation for each ASO service is expected to be mostly disjoint from its SDK-based implementation. When they do arise, however, the cost of the extra work required for a second implementation of a fix will be accepted since the risk will only be present for a few months at most as newer CAPZ releases push SDK-backed service implementations out of support.
+
+Another shorter-term risk is that although ASO v2 is [officially supported by Microsoft](https://azure.github.io/azure-service-operator/introduction/frequently-asked-questions/#what-is-the-support-model), it has not yet published a GA release. Until then, ASO may make breaking changes between pre-releases. The remaining work to be completed before ASO v2's first GA release is being tracked here: https://github.com/Azure/azure-service-operator/milestone/20.
+
+Further down the road, there is an apparent increased risk that CAPZ's lifespan may extend beyond that of its Azure interface when ASO is used instead of the API or SDK directly, as ASO has not yet been proven as a mission-critical interface to Azure to the same extent as Azure's REST API and Go SDK.
+
+Since ASO can be freely installed by users, there is also a risk that a CAPZ-installed ASO may conflict with a user-installed instance. Before CAPZ installs ASO, it should verify that ASO is already either not installed or installed and configured as expected by CAPZ.
```

Comment:
I'm not sure this is a risk, but is rather an implementation detail: when we go to actually implement the standard CAPZ bootstrapping of ASO, we will need to ensure that resource names and namespaces are used in a sane way so as to ~guarantee uniqueness (adding capz-specific prefixes, unique suffixes, etc).

Reply:
I think my main reason for listing this as a risk was because I'm not confident any amount of patching will enable two independent, functional ASO instances to run side-by-side in the same cluster (https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3113#discussion_r1122507238). And given I haven't heard of anyone raising concerns with the current plan, I wonder if it's fair to assume this scenario is unlikely enough to justify not working around the possibility of an existing ASO installation in the immediate-term but at least acknowledge the possibility here.

Reply:
Yes the issue isn't making sure the operator pods, etc are unique. That's easy. The main issue is that both operators will be watching all namespaces by default. The only facility in ASO today control which operator would reconcile which resources is by namespace. By default the operator watches everything in the cluster, so it would be very easy to end up with 2 ASOs each reconciling a resource. Even if we had ownership on the resource that clearly indicated which operator owned it there's still a race to see which operator notices that resource first and picks it up.

Reply:
I see what you mean. Basically we're saying that ASO is designed to be a flexible solution for managing Azure resources on Kubernetes, and if anyone is already using that for any reason in a multi-tenant way w/ Cluster API, then we will probably break their existing multi-tenancy implementation.